e2etest/pkg/executor/executor.go

```diff
@@ -86,27 +92,69 @@ type podDebugExecutor struct {
 	name      string
 	container string
 	image     string
+	ephemeral string
+	cs        clientset.Interface
 }
 
-func ForPodDebug(namespace, name, container, image string) Executor {
-	return &podDebugExecutor{
+func ForPodDebug(cs clientset.Interface, namespace, podName, targetContainer string) (Executor, error) {
+	ret := &podDebugExecutor{
 		namespace: namespace,
-		name:      name,
-		container: container,
-		image:     image,
+		name:      podName,
+		container: targetContainer,
+		ephemeral: "debugger-" + targetContainer,
+		cs:        cs,
+	}
+
+	ephemeralName := "debugger-" + targetContainer
+	ephemeralContainer := corev1.EphemeralContainer{
+		EphemeralContainerCommon: corev1.EphemeralContainerCommon{
+			Name:  ret.ephemeral,
+			Image: agnhostImage,
+		},
+		TargetContainerName: targetContainer,
+	}
+
+	ctx := context.Background()
+	pod, err := cs.CoreV1().Pods(namespace).Get(ctx, podName, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	containerExists := false
+	for _, ec := range pod.Spec.EphemeralContainers {
+		if ec.Name == ret.ephemeral {
+			containerExists = true
+			break
+		}
+	}
+
+	if !containerExists {
+		pod.Spec.EphemeralContainers = append(pod.Spec.EphemeralContainers, ephemeralContainer)
+		_, err = cs.CoreV1().Pods(namespace).UpdateEphemeralContainers(ctx, podName, pod, metav1.UpdateOptions{})
+		if err != nil {
+			return nil, fmt.Errorf("failed to create ephemeral container: %w", err)
+		}
 	}
+
+	return ret, wait.PollUntilContextTimeout(context.Background(), 1*time.Second, 20*time.Second, false, func(context.Context) (bool, error) {
```

Comment:
nit: can the wait be before the return, and the usual if `err != nil` after?

Reply:
don